File: VB/DragAndDropBetweenGridAndTreeList/ImageHelpers/DragTreeListImageHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms
Imports System.Drawing
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Painter
Imports DevExpress.XtraTreeList.ViewInfo
Imports System.Reflection
Imports DevExpress.XtraTreeList.Handler

Namespace DragAndDropBetweenGridAndTreeList
    Public Class DragTreeListImageHelper
        Inherits TreeListPainter

        Private ReadOnly _treeList As TreeList

        Public Sub New(ByVal treeList As TreeList)
            _treeList = treeList
        End Sub

        Public Function GetDragCursor(ByVal hitInfo As TreeListHitInfo, ByVal e As Point) As Cursor
            Dim Handler As PropertyInfo = _treeList.GetType().GetProperty("Handler", BindingFlags.Instance Or BindingFlags.NonPublic)

            Dim handler_Renamed As TreeListHandler = DirectCast(Handler.GetValue(_treeList, Nothing), TreeListHandler)
            Dim fStateData As FieldInfo = handler_Renamed.GetType().GetField("fStateData", BindingFlags.Instance Or BindingFlags.NonPublic)
            Dim data As StateData = DirectCast(fStateData.GetValue(handler_Renamed), StateData)
            Dim info As TreeListViewInfo = _treeList.ViewInfo
            Dim rowInfo As RowInfo = info.GetRowInfoByPoint(hitInfo.MousePoint)
            Dim result As Bitmap = data.GetNodeDragBitmap(rowInfo.Node, rowInfo)
            Dim offset As New Point(rowInfo.Bounds.X, e.Y - rowInfo.Bounds.Y)
            Return CursorCreator.CreateCursor(result, offset)
        End Function
    End Class
End Namespace

File: CS/DragAndDropBetweenGridAndTreeList/ImageHelpers/DragTreeListImageHelper.cs
// Developer Express Code Central Example:
// How to implement drag-and-drop between Grid and TreeList
// 
// This example demonstrates how to create helper classes for implementing
// drag-and-drop between Grid and TreeList. For more information please refer
// to:
// http://www.devexpress.com/scid=A2343
// http://www.devexpress.com/scid=K18056
// http://www.devexpress.com/scid=A1444
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E3375

using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using System.Drawing;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Painter;
using DevExpress.XtraTreeList.ViewInfo;
using System.Reflection;
using DevExpress.XtraTreeList.Handler;

namespace DragAndDropBetweenGridAndTreeList
{
    public class DragTreeListImageHelper : TreeListPainter
    {
        private readonly TreeList _treeList;

        public DragTreeListImageHelper(TreeList treeList)
        {
            _treeList = treeList;
        }

        public Cursor GetDragCursor(TreeListHitInfo hitInfo, Point e)
        {
            PropertyInfo Handler = _treeList.GetType().GetProperty("Handler", BindingFlags.Instance | BindingFlags.NonPublic);
            TreeListHandler handler = (TreeListHandler)Handler.GetValue(_treeList, null);
            FieldInfo fStateData = handler.GetType().GetField("fStateData", BindingFlags.Instance | BindingFlags.NonPublic);
            StateData data = (StateData)fStateData.GetValue(handler);
            TreeListViewInfo info = _treeList.ViewInfo;
            RowInfo rowInfo = info.GetRowInfoByPoint(hitInfo.MousePoint);
            Bitmap result = data.GetNodeDragBitmap(rowInfo.Node, rowInfo.VisibleIndex, rowInfo.IndentInfo.Bounds.Width);
            Point offset = new Point(rowInfo.Bounds.X, e.Y - rowInfo.Bounds.Y);
            return CursorCreator.CreateCursor(result, offset);
        }
    }
}
